### PR TITLE
ci: add mem profiling during functionals run

### PR DESCRIPTION
**Pull request by safchain**
skip skydive-compile-tests
skip skydive-functional-tests-backend-orientdb
skip skydive-go-fmt
skip skydive-k8s-tests
skip skydive-scale-tests
skip skydive-selenium-tests
skip skydive-unit-tests
skip skydive-cdd-overview-tests
skip skydive-ppc64-tests

### CODE REVIEW

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

File: common/profile.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	memory := make(chan os.Signal, 1)
+	signal.Notify(memory, syscall.SIGUSR2)
+
+	for {
+		select {
+		case <-cpu:
+			f, err := os.Create("/tmp/skydive-cpu.prof")
+			if err != nil {
+				log.Fatal("could not create CPU profile: ", err)
+			}
```
Comment (hunchback) [inline common/profile.go]: Missing:

```
defer f.Close()
```

File: common/profile.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	cpu := make(chan os.Signal, 1)
+	signal.Notify(cpu, syscall.SIGUSR1)
+
+	memory := make(chan os.Signal, 1)
+	signal.Notify(memory, syscall.SIGUSR2)
+
+	for {
+		select {
+		case <-cpu:
+			f, err := os.Create("/tmp/skydive-cpu.prof")
```
Comment (hunchback) [inline common/profile.go]: rather use:
```
f, err := ioutil.TempFile("", "skydive-cpu.prof")
```

File: common/profile.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+			log.Println("CPU profiling Started")
+
+			<-cpu
+
+			pprof.StopCPUProfile()
+			log.Println("CPU profiling Stopped")
+		case <-memory:
+			// Memory Profile
+			runtime.GC()
+			memProfile, err := os.Create("/tmp/skydive-memory.prof")
```
Comment (hunchback) [inline common/profile.go]: rather use: `ioutilTempFile`

File: common/profile.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			f, err := os.Create("/tmp/skydive-cpu.prof")
+			if err != nil {
+				log.Fatal("could not create CPU profile: ", err)
+			}
+			if err := pprof.StartCPUProfile(f); err != nil {
+				log.Fatal("could not start CPU profile: ", err)
+			}
+			log.Println("CPU profiling Started")
+
+			<-cpu
```
Comment (hunchback) [inline common/profile.go]: why is this needed?

File: common/profile.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+			pprof.StopCPUProfile()
+			log.Println("CPU profiling Stopped")
+		case <-memory:
+			// Memory Profile
+			runtime.GC()
```
Comment (hunchback) [inline common/profile.go]: this means sending a SIGUSR2 will change the behavior of the application, and the memory dump may not be representative of actual memory typically consumed; if calling `runtime.GC()` explicitly has a favorable effect on the application then it should be invoked periodically by the application regardless of SIGUSR2.

File: scripts/ci/cleanup.sh [hunk 6]
```diff
@@ -69,6 +69,8 @@ EOF
   rm -rf /tmp/skydive_agent* /tmp/skydive-etcd
   rm -rf /var/lib/jenkins/.vagrant.d/tmp
 
+  rm /tmp/skydive-memory.prof*
```
Comment (hunchback) [inline scripts/ci/cleanup.sh]: rather replace by:

```
rm -rf /tmp/skydive-memory.prof*
rm -rf /tmp/skydive-cpu.prof*
```

File: scripts/ci/run-tests-utils.sh [hunk 7]
```diff
@@ -6,6 +6,16 @@ network_setup() {
         done
 }
 
+mem_prof() {
+        echo start memory profiling
+        while(true); do
+                echo trigger memory profiling snapshot
+                sudo pkill -USR2 functionals 2>/dev/null || true
+                sleep 10
+                sudo mv /tmp/skydive-memory.prof /tmp/skydive-memory.prof.$( date +%T ) 2>/dev/null || true
```
Comment (hunchback) [inline scripts/ci/run-tests-utils.sh]: if you create unique file names using `ioutil.TempFile` you can remove this line

**Comment by hunchback** [comment]
@safchain - you probably want to run some checks on the output and *FAIL* the test if cpu/memory is more than should be - alternatively you can simply print the content of these files into the test execution log so that it is kept for later reference.

File: scripts/ci/run-tests-utils.sh (hunk 7 above)
Comment (nplanel) [reply scripts/ci/run-tests-utils.sh]: no, he would like easily keep and found the memory profiling in time order (+%T)

Comment (hunchback) [reply scripts/ci/run-tests-utils.sh]: ok

File: common/profile.go (hunk 5 above)
Comment (safchain) [reply common/profile.go]: The goal here is to see real memory allocation in order to detect memory leak or unexpected memory allocation. If I do not call the GC I will get a lot of noise. In the future we could import this to select what we want to check.

File: common/profile.go (hunk 4 above)
Comment (safchain) [reply common/profile.go]: waiting a second kill to stop the profiling.

**Comment by safchain** [comment]
@hunchback add check on the CI for the kind of thing is a tricky thing that should be address in another PR. The goal here is to add helper to help profiling and troubleshooting

File: common/profile.go (hunk 1 above)
Comment (safchain) [reply common/profile.go]: good catch: missing a close but not a defer

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

File: Makefile [hunk 8]
```diff
@@ -75,6 +75,9 @@ ifeq ($(COVERAGE), true)
   TEST_COVERPROFILE?=../functionals.cover
   EXTRA_ARGS+=-test.coverprofile=${TEST_COVERPROFILE}
 endif
+ifeq ($(WITH_PROF), true)
+  EXTRA_ARGS+= -profile
```
Comment (lebauce) [inline Makefile]: nit: remove leading space
